File: app/src/main/java/com/huhx0015/androidbooster/injections/modules/LocationModule.java
package com.huhx0015.androidbooster.injections.modules;

import android.content.Context;
import android.location.LocationManager;
import androidx.annotation.NonNull;
import com.google.android.gms.location.LocationRequest;
import com.huhx0015.androidbooster.injections.scopes.ServiceScope;
import dagger.Module;
import dagger.Provides;

@Module
public class LocationModule {

    /** CLASS VARIABLES ________________________________________________________________________ **/

    // CONSTANT VARIABLES:
    private static final int UPDATE_INTERVAL = 10000;
    private static final int FASTEST_INTERVAL = 5000;

    /** MODULE METHODS _________________________________________________________________________ **/

    @Provides
    @ServiceScope
    LocationManager providesLocationManager(@NonNull Context context) {
        return (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
    }

    @Provides
    @NonNull
    @ServiceScope
    LocationRequest providesLocationRequest() {
        LocationRequest request = new LocationRequest();
        request.setPriority(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY);
        request.setInterval(UPDATE_INTERVAL);
        request.setFastestInterval(FASTEST_INTERVAL);
        request.setExpirationDuration(UPDATE_INTERVAL);
        return request;
    }
}

File: app/src/main/java/com/huhx0015/androidbooster/injections/modules/LocationModule.kt
package com.huhx0015.androidbooster.injections.modules

import android.content.Context
import android.location.LocationManager
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.Priority
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.components.ServiceComponent
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.android.scopes.ServiceScoped

@Module
@InstallIn(ServiceComponent::class)
object LocationModule {

    /** CONSTANTS ______________________________________________________________________________  */

    private const val UPDATE_INTERVAL = 10000L
    private const val FASTEST_INTERVAL = 5000L

    /** MODULE METHODS _________________________________________________________________________  */

    @Provides
    @ServiceScoped
    fun providesLocationManager(@ApplicationContext context: Context): LocationManager {
        return context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
    }

    @Provides
    @ServiceScoped
    fun providesLocationRequest(): LocationRequest {
        return LocationRequest.Builder(Priority.PRIORITY_BALANCED_POWER_ACCURACY, UPDATE_INTERVAL)
            .setMinUpdateIntervalMillis(FASTEST_INTERVAL)
            .setMaxUpdateDelayMillis(UPDATE_INTERVAL)
            .build()
    }
}